lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -2288,13 +2290,47 @@ func (d *zfs) readonlySnapshot(vol Volume) (string, revert.Hook, error) {
 
 	d.logger.Debug("Created backup snapshot", logger.Ctx{"dev": srcSnapshot})
 
-	// Mount the snapshot directly (not possible through ZFS tools), so that the volume is
-	// already mounted by the time genericVFSBackupVolume tries to mount it below,
-	// thus preventing it from trying to unmount it at the end, as this is a custom snapshot,
-	// the normal mount and unmount logic will fail.
-	err = TryMount(srcSnapshot, tmpDir, "zfs", 0, "")
-	if err != nil {
-		return "", nil, err
+	if d.isBlockBacking(vol) {
+		// For block devices, we make them appear.
+		// Check if already active.
+		current, err := d.getDatasetProperty(dataset, "volmode")
+		if err != nil {
+			return "", nil, err
+		}
+
+		if current != "dev" {
+			// Activate.
+			err = d.setDatasetProperties(dataset, "volmode=dev", "snapdev=visible")
+			if err != nil {
+				return "", nil, err
+			}
+
+			defer func() { _ = d.setDatasetProperties(dataset, "volmode=none", "snapdev=hidden") }()
+
+			// Wait a second to give udev a chance to kick in.
+			time.Sleep(time.Second)
+
+			d.logger.Debug("Activated ZFS volume", logger.Ctx{"dev": dataset})
+		}
+
+		volPath, err := d.getVolumeDiskPathFromDataset(srcSnapshot)
+		if err != nil {
+			return "", nil, err
+		}
+
+		err = TryMount(volPath, tmpDir, vol.ConfigBlockFilesystem(), unix.MS_RDONLY, "")
```

Comment:
Its interesting you've added MS_RDONLY here, I'm thinking perhaps we also need this for the `else` case below too? What do you think?

Reply:
IIRC XFS complained when I omitted `MS_RDONLY`. The `else` case doesn't seem to complain.

Reply:
Yeah I'm not saying its wrong, I thought perhaps we could benefit from doing it in the else stanza do as its specifically supposed to be a read only mount.

Reply:
To reinforce its read only.